### PR TITLE
NetworkMgr: Refine isConnected check

### PR DESCRIPTION
**Pull request by NiLuJe**
Spent some time testing #10062 today, and there were a few shortcomings ;).

* `carrier` is set to 1 as soon as the device is *administratively* up (in practice, as soon as we run `ifconfig up`). This is perfectly fine for `isWifiOn`, but absolutely not for `isConnected`, because we are not, actually, connected to *anything*, no attempt at associating has even been made at that point. Besides being semantically wrong, in practice, this will horribly break the connectivity check, because it expects that `isConnected` means we can talk to at least the LAN.
* Delving into the Linux docs reveals that `operstate` looks like a better candidate, as it reflects *operational status*; for Wi-Fi, that means associated and successfully authenticated. That's... closer, but still not it, because we still don't have an IP, so we technically can't talk to anything other than the AP.
* So, I've brought out the big guns (`getifaddrs`), and replicated a bit of code that I already use in the USBNetwork hack on Kindle, to detect whether we actually have an IP assigned. (Other approaches, like `/proc/net/route`, may not be entirely fool-proof, and/or get complicated when IPv6 enters the fray (which it does, on Kobo, Mk. 8+ devices are IPv6-enabled)).

TL;DR: Bunch of C via ffi, and `isConnected` now returns true only when the device is operationally up *and* we have an IP assigned.

I've only tested this on Kobo right now, so ping @yparitcher to check how it fares on Kindle ;).

Depends on https://github.com/koreader/koreader-base/pull/1579 & https://github.com/koreader/lj-wpaclient/pull/10

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/10098)

### CODE REVIEW

**Comment by NiLuJe** [comment]
Sidebar for testers: I've only swapped *Kobo* to the new check, so you'll have to fudge your platform's `initNetworkManager` implementation in `device.lua` first ;).

**Comment by NiLuJe** [comment]
Another question: should the `Network info` menu entry be locked behind `isConnected` or is`WifiOn`? `isConnected` makes sense, as it makes a gateway ping, so it would be pretty useless in the intermediary state, but I vaguely recall having weird requirements about it in some setups...

**Comment by Frenzie** [comment]
> Another question: should the Network info menu entry be locked behind isConnected or isWifiOn? isConnected makes sense, as network info makes a gateway ping, so it would be pretty useless in the intermediary state, but I vaguely recall people having weird requirements about it in some setups...

In its present form I'd think it should be behind isConnected?

**Comment by yparitcher** [comment]
Looks good to my untrained eye, sorry to bring this mess upon us ;)

I will test on kindle, probably not before sunday.

File: frontend/device/kobo/device.lua [hunk 1]
```diff
@@ -829,10 +829,7 @@ function Kobo:initNetworkManager(NetworkMgr)
         self:reconnectOrShowNetworkMenu(complete_callback)
     end
 
-    local net_if = os.getenv("INTERFACE")
-    if not net_if then
-        net_if = "eth0"
-    end
+    local net_if = os.getenv("INTERFACE") or "eth0"
```
Comment (Frenzie) [inline frontend/device/kobo/device.lua:832]: On a not quite related note (because not applicable to Kobo or whatever), I use this in a local shell script:

```sh
nmcli -t -f device connection show | head -n1
```

**Comment by NiLuJe** [comment]
> Looks good to my untrained eye, sorry to bring this mess upon us ;)

Eh, no worries, it was sorely in need of a good cleanup ;).

**Comment by NiLuJe** [comment]
> In its present form I'd think it should be behind isConnected?

Agreed. We're good then, as that's what it's behind since #10062 ;).

(Note to self: drop the IP logging before merge; it's fun, but chatty; and this check means you'll see it twice when connecting via the menu: once from the connectivity check, and once from the menu refresh).

File: frontend/device/kobo/device.lua (hunk 1 above)
Comment (NiLuJe) [reply frontend/device/kobo/device.lua:832]: In a similarly related tangent, the "nice" way of detecting IP assignment would be to *listen* to the netlink socket (`getifaddrs` uses it to pull the data it offers); something that is usually massaged into a slightly more palatable form over dbus via NM ;). (I'm told: I don't use NM :D).

**Review by yparitcher — APPROVED** [review]
It seems to work on my kindle.
I did not do any complicated testing, just turning the wifi on & off an making sure the menu & network info updated accordingly.

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 1 of 2 files at r1, 2 of 3 files at r2, 4 of 4 files at r4, 2 of 2 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/10098)* status: all files reviewed, 1 unresolved discussion (waiting on @Frenzie)
